lndclient/router_client.go

```diff
@@ -226,17 +231,27 @@ func (r *routerClient) trackPayment(ctx context.Context,
 		for {
 			payment, err := stream.Recv()
 			if err != nil {
-				switch status.Convert(err).Code() {
-
+				code := status.Convert(err).Code()
+				switch {
 				// NotFound is only expected as a response to
 				// TrackPayment.
-				case codes.NotFound:
+				case code == codes.NotFound:
 					err = channeldb.ErrPaymentNotInitiated
 
 				// NotFound is only expected as a response to
 				// SendPayment.
-				case codes.AlreadyExists:
+				case code == codes.AlreadyExists:
 					err = channeldb.ErrAlreadyPaid
+
+				// If the stream has exited on the server side,
+				// which may be the case when the payment has
+				// reached a final state, we return a stream
+				// closed error so that callers can handle this
+				// case specifically.
+				case code == codes.Unknown &&
+					err.Error() == "EOF":
+
+					err = ErrStreamComplete
```

Comment:
If we close both channels, we can't keep listening on the error channel in a select once the stream is done, select on a closed channel will always trigger. I think it's preferable to leave the behavior as is, and add a special case so that we can identify this case when we want to.

Reply:
Ok good point about bigger selects. And how about not sending anything at all? The success or failure result is already the EOF marker really.

Reply:
That would be a behaviour change for other calls using this endpoint (although I don't know how much it matters, since the update would always arrive first). Could also move the `EOF` check into our loop which has this specific use case and leave the client untouched?

Reply:
Behavior change is also a valid point. It just feels a bit strange to always end the happy flow with an error. Close channel  to me is the most natural way to map the grpc stream close still. Another alternative for bigger selects is to set the status channel to nil once you've read a close, so that the event loop can keep going.

Reply:
> It just feels a bit strange to always end the happy flow with an error

This is already the case with the EOF if you keep listening after settle? This just makes it an error that we specifically match. Thinking more about this I think we should just catch it in the caller, since this is a very specific case. 

> set the status channel to nil once you've read a close

This feels like it would be racey? We replace the channel with nil at the same time we try to read from it? It won't be an issue in practice, but I suspect it will trigger unit test race detection.

Reply:
>This is already the case with the EOF if you keep listening after settle?

If the channel is closed, you will at least get a `!ok` in `update, ok := <- updateChan`. It maps better to go channels than a custom err?

> We replace the channel with nil at the same time we try to read from it?

I mean replace the channel with `nil` in the caller select. So:
```
case update, ok := <- updateChan:
   if !ok {
      updateChan = nil
      continue
   }
````

Reply:
Ok yeah nice, I thought you meant in the client itself.